File: arduino/pcc2arduino.cpp
#include "pcc2arduino.h"

///////////////////////
// reads in bytes from serial
// returns true when the controller state is ready
///////////////////////
bool PCController::update()
{
	if (!ser || !ser->available())
		return false;
	
	//find the start of a valid parsel
	while (ser->peek() != '<')
	{
		byte b = ser->read();
		
		if (b == -1)
			return false;
	}	
	byte parsel[10];
	ser->readBytes(parsel, 10);
	
	if (parsel[0] == '<' && parsel[9] == '>')
	{
		//copies the data part of the parsel into the input, to be parsed
		memcpy(inp, parsel + 1, 8); 
		
		parseButtons();
		parseAxes();
		
		ser->write('*');
		
		return true;
	}
	
	return false;
}


bool PCController::updateInc()
{
	static int i = 0;
	static bool reading = false;
	
	if(!ser)
		return false;
	
	//Notify the system
	ser->write('*');
	
	byte b;
	
	//wait for response
	while(!ser->available());
	b = ser->read();
	
	if (i == 0 && b == '<')
	{
		reading = true;
		return false;
	}
	  
	if (reading && i >= 0 && i < 8)
	{
		inp[i] = b;
		i++;
		return false;
	}
	
	if (reading && i == 8 && b == '>')
	{
		reading = false;
		i = 0;
		
		parseButtons();
		parseAxes();
		
		return true;
	}
	
	return false;
}
/////////////////////////////////////////////////
// Parses 2 bytes into 16 buttons (bools)
/////////////////////////////////////////////////
void PCController::parseButtons()
{
	dpad_up    = inp[0] & 1;
	dpad_down  = inp[0] & 2;
	dpad_left  = inp[0] & 4;
	dpad_right = inp[0] & 8;
		
	a = inp[0] & 16;
	b = inp[0] & 32;
	x = inp[0] & 64;
	y = inp[0] & 128;
		
	lb = inp[1] & 1;
	rb = inp[1] & 2;
	
	lt = inp[1] & 4;
	rt = inp[1] & 8;
	
	select = inp[1] & 16;
	start  = inp[1] & 32;
	
	ls_press = inp[1] & 64;
	rs_press = inp[1] & 128;
}

////////////////////////////////////////////////
// Parses 12 bytes into 6 axes, 2 bytes per axis
// Serial architecture is "little-endian"
// UPDATE: axis values are now 0-255, one byte
//         ... makes conversion a lot simpler
////////////////////////////////////////////////
void PCController::parseAxes()
{
	x_axis = inp[2];
	
	y_axis = inp[3];
	
	z_axis = inp[4];
	
	r_axis = inp[5];
	
	u_axis = inp[6];
	
	v_axis = inp[7];
}

void PCController::displayButtons()
{
	if (ser)
	{
		ser->print("up:    ");
		ser->print(dpad_up);
		ser->println();
		ser->print("down:  ");
		ser->print(dpad_down);
		ser->println();
		ser->print("left:  ");
		ser->print(dpad_left);
		ser->println();
		ser->print("right: ");
		ser->print(dpad_right);
		ser->println();

		ser->print("a: ");
		ser->print(a);
		ser->println();
		ser->print("b: ");
		ser->print(b);
		ser->println();
		ser->print("x: ");
		ser->print(x);
		ser->println();
		ser->print("y: ");
		ser->print(y);
		ser->println();

		ser->print("lb: ");
		ser->print(lb);
		ser->println();
		ser->print("rb: ");
		ser->print(rb);
		ser->println();
		ser->print("lt: ");
		ser->print(lt);
		ser->println();
		ser->print("rt: ");
		ser->print(rt);
		ser->println();

		ser->print("select: ");
		ser->print(select);
		ser->println();
		ser->print("start:  ");
		ser->print(start);
		ser->println();
	
		ser->print("LS press: ");
		ser->print(ls_press);
		ser->println();
		ser->print("RS press: ");
		ser->print(rs_press);
		ser->println();
		ser->println();
	}
}
void PCController::displayAxes()
{
	if (ser)
	{
		ser->print("X axis: ");
		ser->print(x_axis);
		ser->println();

		ser->print("Y axis: ");
		ser->print(y_axis);
		ser->println();

		ser->print("Z axis: ");
		ser->print(z_axis);
		ser->println();

		ser->print("R axis: ");
		ser->print(r_axis);
		ser->println();

		ser->print("U axis: ");
		ser->print(u_axis);
		ser->println();

		ser->print("V axis: ");
		ser->print(v_axis);
		ser->println();
		ser->println();
	}
}

File: pc/pcc2arduino.py
import serial
import pygame
import sys
from time import sleep

def mapaxis(ax):
        return int(ax * 127 + 128)
class joy_grabber:
    bn = {'a': 0, 'b': 1, 'x': 2, 'y': 3, 'lb': 4, 'rb': 5, 'select': 6, 'start': 7, 'ls': 8, 'rs': 9}
    def __init__(self, id):
        self.joy = pygame.joystick.Joystick(id)
        self.joy.init()

        self.state = bytearray(10)
    
    def update(self):
        self.state[0] = ord('<')

        ###############################
        # FIRST BYTE
        ###############################
        b1 = 0
        if self.joy.get_hat(0)[1] == 1: # dpad_up
            b1 += 1
        elif self.joy.get_hat(0)[1] == -1: # dpad_down
            b1 += 2

        if self.joy.get_hat(0)[0] == -1: # dpad_left
            b1 += 4
        elif self.joy.get_hat(0)[0] == 1: # dpad_right
            b1 += 8

        b1 += self.joy.get_button(0) * 16  #a
        b1 += self.joy.get_button(1) * 32  #b
        b1 += self.joy.get_button(2) * 64  #x
        b1 += self.joy.get_button(3) * 128 #y

        self.state[1] = b1

        ###############################
        # SECOND BYTE
        ###############################
        b2 = 0

        b2 += self.joy.get_button(4) * 1 #lb
        b2 += self.joy.get_button(5) * 2 #rb

        if sys.platform == 'win32':
            if self.joy.get_axis(2) > 0.9: #lt
                b2 += 4
            elif self.joy.get_axis(2) < -0.9: #rt
                b2 += 8
        else:
            if self.joy.get_axis(2) > 0.9:
                b2 += 4
            if self.joy.get_axis(5) > 0.9:
                b2 += 8

        b2 += self.joy.get_button(6) * 16  #select
        b2 += self.joy.get_button(7) * 32  #start
        b2 += self.joy.get_button(8) * 64  #ls_press
        b2 += self.joy.get_button(9) * 128 #rs_press

        self.state[2] = b2

        ################################
        # Left and right sticks
        ################################
        self.state[3] = mapaxis(self.joy.get_axis(0)) #x_axis
        self.state[4] = mapaxis(self.joy.get_axis(1)) #y_axis

        if sys.platform == 'win32':
            trig = self.joy.get_axis(2)
            if trig > 0:
                self.state[5] = int(trig * 128)  #z_axis
            else:
                self.state[6] = int(trig * -128) #r_axis
        else:
            self.state[5] = mapaxis(self.joy.get_axis(2))
            self.state[6] = mapaxis(self.joy.get_axis(5))

        self.state[7] = mapaxis(self.joy.get_axis(3)) #u_axis
        self.state[8] = mapaxis(self.joy.get_axis(4)) #v_axis

        self.state[9] = ord('>')
    
    def send(self, ser):
        ser.write(self.state)
        ser.read()

    def kill(self):
        return self.joy.get_button(6) and self.joy.get_button(7)


# Code to determine serial port
port = ''
if sys.platform == 'win32':
    port = 'COM3'
elif sys.platform == 'linux':
    port = '/dev/ttyACM0'

if len(sys.argv) >= 2:
    port = sys.argv[1]


ser = serial.Serial(port, baudrate=115200)
sleep(2.5)

pygame.init()
pygame.joystick.init()

js = joy_grabber(0)

print('Reading joystick data. Press select and start to exit.')
while not js.kill():
    pygame.event.get()

    js.update()
    js.send(ser)
        
pygame.joystick.quit()
pygame.quit()
